File: core/core-api/src/main/java/com/coffee/gu/support/pg/PaymentRecoverScheduler.java
package com.coffee.gu.support.pg;

import com.coffee.gu.PGConfirmResult;
import com.coffee.gu.PGPayment;
import com.coffee.gu.PaymentGatewayStatus;
import com.coffee.gu.cancel.CancelService;
import com.coffee.gu.order.Order;
import com.coffee.gu.order.OrderReader;
import com.coffee.gu.payment.Payment;
import com.coffee.gu.payment.PaymentCompleter;
import com.coffee.gu.payment.PaymentGatewayProcessor;
import com.coffee.gu.payment.PaymentManager;
import com.coffee.gu.payment.PaymentReader;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.OffsetDateTime;

@Component
public class PaymentRecoverScheduler {
    private static final int MAX_RETRY_COUNT = 5;
    public static final int LIMIT = 20;
    private final PaymentReader paymentReader;
    private final OrderReader orderReader;
    private final PaymentGatewayProcessor paymentGatewayProcessor;
    private final PaymentCompleter paymentCompleter;
    private final PaymentManager paymentManager;
    private final CancelService cancelService;

    public PaymentRecoverScheduler(PaymentReader paymentReader, OrderReader orderReader, PaymentGatewayProcessor paymentGatewayProcessor, PaymentCompleter paymentCompleter, PaymentManager paymentManager, CancelService cancelService) {
        this.paymentReader = paymentReader;
        this.orderReader = orderReader;
        this.paymentGatewayProcessor = paymentGatewayProcessor;
        this.paymentCompleter = paymentCompleter;
        this.paymentManager = paymentManager;
        this.cancelService = cancelService;
    }

    @Scheduled(cron = "0 * * * * *")
    @Transactional
    public void schedule() {
        for (Payment pendingPayment : paymentReader.getPendingPayments(LIMIT)) {
            Order order = orderReader.getByOrderKey(pendingPayment.getOrderKey());

            if (pendingPayment.isExpired(Duration.ofMinutes(30))) {
                try {
                    cancelService.cancel(order);
                } catch (Exception e) {
                    pendingPayment.fail();
                    paymentManager.save(pendingPayment);
                }
                continue;
            }

            try {
                PGPayment pgPayment = paymentGatewayProcessor.getPGPayment(pendingPayment.getOrderKey());
                if (pgPayment.status() == PaymentGatewayStatus.DONE) {
                    PGConfirmResult pgConfirmResult = PGConfirmResult.success(
                            pgPayment.orderKey(),
                            pgPayment.paymentKey(),
                            pendingPayment.getMethod(),
                            pendingPayment.getApproveCode(),
                            pendingPayment.getPaidAt() != null ? pendingPayment.getPaidAt() : OffsetDateTime.now()
                    );
                    paymentCompleter.complete(order, pendingPayment.getId(), pgConfirmResult);
                } else {
                    cancelService.cancel(order);
                }
            } catch (Exception e) {
                pendingPayment.increaseRetryCount();
                if (pendingPayment.isRetryLimitExceeded(MAX_RETRY_COUNT)) {
                    pendingPayment.fail();
                    // 알림
                }
                paymentManager.save(pendingPayment);
            }
        }
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/support/pg/PaymentRecoverScheduler.kt
package com.coffee.gu.support.pg

import com.coffee.gu.PGConfirmResult
import com.coffee.gu.PaymentGatewayStatus
import com.coffee.gu.cancel.CancelService
import com.coffee.gu.enums.PaymentMethod
import com.coffee.gu.order.OrderReader
import com.coffee.gu.payment.PaymentCompleter
import com.coffee.gu.payment.PaymentGatewayProcessor
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.Duration
import java.time.OffsetDateTime

@Component
class PaymentRecoverScheduler(
    private val paymentRecoveryProcessor: PaymentRecoveryProcessor,
    private val orderReader: OrderReader,
    private val paymentGatewayProcessor: PaymentGatewayProcessor,
    private val paymentCompleter: PaymentCompleter,
    private val cancelService: CancelService,
) {
    companion object {
        private const val MAX_RETRY_COUNT = 5
        const val LIMIT = 20
    }

    @Scheduled(cron = "0 * * * * *")
    fun schedule() {
        val pendingPayments = paymentRecoveryProcessor.claimPendingPayments(LIMIT)
        for (pendingPayment in pendingPayments) {
            val order = orderReader.getByOrderKey(pendingPayment.orderKey)

            if (pendingPayment.isExpired(Duration.ofMinutes(30))) {
                try {
                    cancelService.cancel(order)
                } catch (e: Exception) {
                    paymentRecoveryProcessor.handleExpireFail(pendingPayment)
                }
                continue
            }

            try {
                val pgPayment = paymentGatewayProcessor.getPGPayment(pendingPayment.orderKey)
                if (pgPayment.status == PaymentGatewayStatus.DONE) {
                    val pgConfirmResult = PGConfirmResult.success(
                        pgPayment.orderKey,
                        pgPayment.paymentKey,
                        pendingPayment.method ?: PaymentMethod.CARD,
                        pendingPayment.approveCode,
                        pendingPayment.paidAt ?: OffsetDateTime.now()
                    )
                    paymentCompleter.complete(order, pendingPayment.id, pgConfirmResult)
                } else {
                    cancelService.cancel(order)
                }
            } catch (e: Exception) {
                paymentRecoveryProcessor.handleRetry(pendingPayment, MAX_RETRY_COUNT)
            }
        }
    }
}
